### PR TITLE
Various changes and features (revision 2)

### PR DESCRIPTION
**Pull request by pxlvxl**
- added `/:paletteSlug/:resolution` functionality for localhost testing
- created `currFile.sublayers` for *things that should zoom with the canvas layers*
- `currFile.layers` now solely contains the canvas layers
- added `getProjectData` to `FileManager`'s exported methods
---
- added `FileManager.localStorageSave` (it's basically just: localStorage.setItem("lpe-cache",FileManager.getProjectData()))
- added `FileManager.localStorageCheck` (it's basically just: `!!localStorage.getItem("lpe-cache")`)
- added `FileManager.localStorageLoad` (it's basically just: `return localStorage.getItem("lpe-cache")`)
- added `FileManager.localStorageReset` (for debugging purity)
---
- calling `FileManager.localStorageSave()` on mouse up (we should stress test this)
---
- changed lpe file format to `{canvasWidth:number,canvasHeight:number,selectedLayer:number,colors:[],layers:[]}`
- added backward compatibility for the old lpe file format
---
- added some canvas utility functions in `canvas_util`
- added Unsettled's color similarity utility functions in `color_util2`
---
- html boilerplate - wang tiles
- POC - tiny text boilerplate 
- POC - tiny text font scraper
---
- WIP - added two optional url route parameters `/:paletteSlug/:resolution/:rowWidth/:binStr`
- WIP POC - hbs_parser.js (outputs tree data about hbs file relationships)

### CODE REVIEW

File: js/Events.js [hunk 1]
```diff
@@ -100,8 +100,9 @@ const Events = (() => {
      */
      function on(event, elementId, functionCallback, ...args) {
         //if element provided is string, get the actual element
+        if(!elementId)return;
```
Comment (unsettledgames) [inline js/Events.js:103]: Let getElement take care of this imo, it would create less clutter when calling getElement

File: js/File.js [hunk 2]
```diff
@@ -2,11 +2,12 @@ class File {
     // Canvas, canvas state
     canvasSize = [];
     zoom = 7;
-    canvasView = document.getElementById("canvas-view");
+    canvasView = document.getElementById("canvas-view") ?? document.createElement("canvas");
```
Comment (unsettledgames) [inline js/File.js:5]: canvasView is a div, not a canvas, but other than that I'd avoid error checking on DOM elements, creating an empty element only delays the inevitable in my opinion

File: js/File.js [hunk 3]
```diff
@@ -246,7 +247,7 @@ class File {
         this.rcPivot = "topleft";
 
         // Computing the min and max coordinates in which there's a non empty pixel
-        for (let i=1; i<currFile.layers.length - nAppLayers; i++) {
+        for (let i=0; i<currFile.layers.length; i++) {
```
Comment (unsettledgames) [inline js/File.js:250]: Being finally able to do this is sooo good

File: js/FileManager.js [hunk 4]
```diff
@@ -55,6 +57,9 @@ const FileManager = (() => {
 
         if (typeof ga !== 'undefined')
             ga('send', 'event', 'Pixel Editor Save', selectedPalette, currFile.canvasSize[0]+'/'+currFile.canvasSize[1]); /*global ga*/
+
+        
+        LayerList.closeOptionsMenu(); // is this the right place for this?
```
Comment (unsettledgames) [inline js/FileManager.js:62]: I think it's unnecessary, it just closes the context menu for the layers, but at this point it should already be closed by all the other mouse events

File: js/LayerList.js [hunk 5]
```diff
@@ -97,7 +98,7 @@ const LayerList = (() => {
                 toMergeImageData.data[i+2], toMergeImageData.data[i+3]
             ];
 
-            let currentUnderlyingPixel = [
+            let currentUnderlyingPixel = [ //TODO: I'd be curious to know if this array slows this function down
```
Comment (unsettledgames) [inline js/LayerList.js:101]: For big sprites, probably. currentUnderlyingPixel shouldn't be modified iirc, so we can just use belowImageData.data I guess

File: js/TopMenuModule.js [hunk 6]
```diff
@@ -1,7 +1,7 @@
 const TopMenuModule = (() => {
 
-    const mainMenuItems = document.getElementById('main-menu').children;
-    let infoList = document.getElementById('editor-info');
+    const mainMenuItems = document.getElementById('main-menu')?.children ?? [];
```
Comment (unsettledgames) [inline js/TopMenuModule.js:3]: Same as the thing with the canvasView, if the elements are not there, we're only delaying the inevitable (same at line 4)

File: js/color_utils3.js [hunk 7]
```diff
@@ -0,0 +1,338 @@
...
+        "#ffa214",
+        "#ffc825",
+        "#0098dc",
+        "#00cdf9",
+        "#0cf1ff",
+        "#7a09fa",
+        "#3003d9"
+    ]
+};
+const COLOR_META = {
```
Comment (unsettledgames) [inline js/color_utils3.js:86]: Can you comment this a bit please? Up to line 135

**Comment by lospec** [comment]
seems to be a z-index issue

image

**Pull request merged**